aodntools/timeseries_products/common.py

```diff
@@ -6,6 +6,50 @@ class NoInputFilesError(Exception):
     pass
 
 
+def get_qc_variable_names(nc):
+    """
+    Return a list of the variables in a file with names ending in '_quality_control'
+
+    :param nc: open xarray Dataset
+    :return: list of variable names
+    """
+    varlist = list(nc.variables)
+    return [v for v in varlist if v.endswith('_quality_control')]
+
+def check_imos_flag_conventions(nc, varnames=None):
+    """
+    Check that the given QC variables are using the IMOS flag conventions (according to the
+    quality_control_conventions variable attribute).
+    If no variable names given, check all variables with names ending in '_quality_control'.
+    Returns a list of error messages (empty if all ok).
+
+    :param nc: open xarray Dataset
+    :param varnames: list of variables to check, or None to check all QC variables
+    :return: list of error messages
+    """
+    if varnames is None:
+        varnames = get_qc_variable_names(nc)
+    if isinstance(varnames, str):
+        varnames = [varnames]
+
+    # accept two variants on the convention name, used in versions 1.3 and 1.4 of the
+    # IMOS NetCDF Conventions document
+    accepted_conventions = {"IMOS standard flags", "IMOS standard set using the IODE flags"}
+    errors = set()
+    for var in varnames:
+        if var not in nc.variables:
+            errors.add('variable {var} not in file'.format(var=var))
+            continue
+        conventions = getattr(nc[var], 'quality_control_conventions', None)
+        if conventions is None:
+            errors.add('variable {var} missing quality_control_conventions'.format(var=var))
+            continue
+        if conventions not in accepted_conventions:
+            errors.add('unexpected quality_control_conventions: "{conventions}"'.format(conventions=conventions))
+
+    return sorted(errors)
```

Comment:
Maybe is better to sort the varnames instead and use a list for errors, as is done elsewhere.

Reply:
Yes, but your msg will be out of order because of the set. In the error event, I would expect to see errors in a per variable fashion and in order of importance (variable not in file > missing_qc > general conventions). Using a list and moving the accepted_conventions out of the loop solve the out of order and  over-reporting problem!